resources/type-system/upper-lower-bounds.md

```diff
@@ -81,6 +85,34 @@ We define the upper bound of two types T1 and T2 to be **UP**(`T1`,`T2`) as foll
 - **UP**(`T1`, `T2`) = `T2` if **BOTTOM**(`T1`)
 - **UP**(`T1`, `T2`) = `T1` if **BOTTOM**(`T2`)
 
+- **UP**(`X1 extends B1`, `T2`) =
+  - `T2` if `X1 <: T2`
+  - otherwise `X1` if `T2 <: X1`
+  - otherwise **UP**(`B1a`, `T2`)
+    where `B1a` is the greatest closure of `B1` with respect to `X1`,
+    as defined in [inference.md].
+
+- **UP**(`X1 & B1`, `T2`) =
+  - `T2` if `X1 <: T2`
+  - otherwise `X1` if `T2 <: X1`
+  - otherwise **UP**(`B1a`, `T2`)
+    where `B1a` is the greatest closure of `B1` with respect to `X1`,
+    as defined in [inference.md].
+
+- **UP**(`T1`, `X2 extends B2`) =
+  - `X2` if `T1 <: X2`
+  - otherwise `T1` if `X2 <: T1`
+  - otherwise **UP**(`T1`, `B2a`)
+    where `B2a` is the greatest closure of `B2` with respect to `X2`,
+    as defined in [inference.md].
+
+- **UP**(`T1`, `X2 & B2`) =
+  - `X2` if `T1 <: X2`
+  - otherwise `T1` if `X2 <: T1`
+  - otherwise **UP**(`T1`, `B2a`)
+    where `B2a` is the greatest closure of `B2` with respect to `X2`,
+    as defined in [inference.md].
```

Comment:
Could we have a rule of
- **UP**(`X & B1`, `X & B2`) =
  - `X & B3` if `B3` <: `B0`
  - `X` otherwise
  where `B3` = **UP**(`B1`, `B2`), `B0` is the bound of `X`.

With the rules above, we always end up at the `X` in that case.

Reply:
Right, I think we could have that. It's probably not going to be used very often, but it seems possible, and it would improve on the typing in those situations where it is applicable.

Reply:
Maybe, some day.